torch/csrc/jit/passes/freeze_module.cpp

```diff
@@ -318,55 +291,151 @@ class AttributePropagator {
     }
   }
 
-  // cleanupFrozenModule function cleans up the Frozen module. it performs the
+  bool moduleEscapes(Module& subModule, std::shared_ptr<Graph>& graph) {
+    for (auto& output : graph->outputs()) {
+      if (output->type() == subModule.type()) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  // cleanupFrozenModule function cleans up the Frozen module. It performs the
   // following:
   // 1) Remove unused attributes.
   // 2) Remove unreferenced submodules
   // 3) Remove non public unreferenced methods.
-  // TODO: do #3 because there is no API to 'unsafely' remove methods.
   void cleanupFrozenModule(std::shared_ptr<Graph>& graph) {
-    std::vector<std::string> attrsToRemove;
-    auto type = module_.type();
-    size_t N = type->numAttributes();
     recordReferencedAttrs(graph);
-    auto it = preservedAttrs_.find(module_._ivalue());
-    auto it2 = preservedScalarAttrs_.find(module_._ivalue());
+    handleSharedClassType(module_, graph);
+    removeUnusedAttrs();
+  }
+
+  // Prepraring for clean up phase. At this point, record all  subModules that
+  // contains mutable attributes.
+  void recordReferencedAttrs(std::shared_ptr<Graph>& graph) {
```

Comment:
How has this function changed? it's hard to tell because the code got reorganized.

Reply:
I had to change it. Because initial implementation checks only toplevel module. Now it is extended to scan submodules.